tensorflow_model_optimization/g3doc/guide/clustering/clustering_example.ipynb

```diff
@@ -0,0 +1,686 @@
+{
```

Comment:
<p>Please see the pruning example with how it doesn't use tf.nn.softmax on the last layer directly and instead SparseCategoricalCrossEntropy with from_logits. </p><p> </p><p>The reasoning is that it's not numerically stable to compute log (softmax(..)) directly, as described in places such as&nbsp;<a href="https://blog.feedly.com/tricks-of-the-trade-logsumexp/" target="_blank">https://blog.feedly.com/tricks-of-the-trade-logsumexp/</a>&nbsp;. The log part is in the `sparse_categorical_crossentropy' loss, which operates</p><p>on the softmax from tf.nn.softmax. </p><p> </p><p>SparseCategoricalCrossEntropy(from_logits=True) can use the numerically stable code. In the current code with softmax and log in separate functions, it's hard to always use the numerically stable part.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/model-optimization/pull/330/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Reply:
<p>Thanks! This is interesting.</p><p>By&nbsp;overcoming&nbsp;numerical&nbsp;instability, do you mean we need to stabilize the&nbsp;softmax&nbsp;function against underflow and overflow?&nbsp;If yes, I agree that we need to use the&nbsp;<a href="https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/backend.py#L4652" target="_blank">tensorflow&nbsp;implementation</a>.</p><p>Side questions:</p><p>1) If we use&nbsp;SparseCategoricalCrossEntropy&nbsp;with&nbsp;from_logits=false combined with the last layer of the model as&nbsp;softmax, it will end up with this <a href="https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/backend.py#L4621" target="_blank">clipping</a> in Keras, which generates a bound [1e-7, 1-1e-7] on the output probability. However, will this clipping work if the input values are NaNs already? It should be no? But if yes, is the bound keeping the output of log(softmax(...)) stable?</p><p>2) In general, what use cases will be suited for the Keras clipping?</p>

Reply:
re 1): The probability calculation never generates nans. But it can saturate to `0.0` or `1.0`, which will generate nans in the loss calculation, so clipping prevents the nans from being generated.  

Also note that in graph mode keras can bypasses the softmax and executes the `from_logits=True` path, and so the results can be **different** depending on **how you're using the model**. That's bad.

re 2): None. It keeps beginners out of trouble, but for any serious application you shouldn't use it. Any example that hits the saturation gets a gradient of zero => is being **ignored** by the training procedure. For a small number of classes this will be all your worst classified examples (which would be an important signal). For a large number of classes this could be a significant fraction of your data.

If you want to export a model that outputs probabillities, add a softmax layer immediately before exporting.

Reply:
<blockquote>&nbsp;The probability calculation never generates nans</blockquote><p>What is the probability here?</p><p>Nans can be generated after softmax function (when the denominator of the softmax becomes 0), which is the input for keras <a href="https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/backend.py#L4621" target="_blank">clipping</a> afaik.</p><p> </p><blockquote>Also note that in graph mode keras can bypasses the softmax and executes the&nbsp;<code>from_logits=True</code>&nbsp;path,</blockquote><p>Did not know that. Will look for the code at some point. Thanks!</p><p> </p><blockquote>re 2):&nbsp;</blockquote><p>Agreed. Thanks!</p>

Reply:
> What is the probability here?

The softmax output.

>  denominator of the softmax becomes 0

How do you make the denominator 0? it't the sum of exponentials. Sure some of those exponentials could underflow, but any good implementations subtracts the largest logit from all the logits (`logits = logits - max(logits)`) So it's impossible for _everything_ to underflow.... unless all your logits are -inf? If you have infs or nans in your logits then you have a different problem.

Right?

Reply:
Thanks!
> How do you make the denominator 0? 
I was thinking in exp(x) when x is very very negative (not sure whether it has to be -inf), the exp(x) will underflow.

> Sure some of those exponentials could underflow, but any good implementations subtracts the largest logit from all the logits
Yes, indeed. And I did not assume all implementation already has this. Will check the details myself.